File: mobile/library/swift/src/Headers.swift
import Foundation

/// Base class that is used to represent header/trailer data structures.
/// To instantiate new instances, see `{Request|Response}HeadersBuilder`.
@objcMembers
public class Headers: NSObject {
  let headers: [String: [String]]

  /// Get the value for the provided header name.
  ///
  /// - parameter name: Header name for which to get the current value.
  ///
  /// - returns: The current headers specified for the provided name.
  public func value(forName name: String) -> [String]? {
    return self.headers[name]
  }

  /// Internal initializer used by builders.
  ///
  /// - parameter headers: Headers to set.
  required init(headers: [String: [String]]) {
    self.headers = headers
    super.init()
  }
}

// MARK: - Equatable

extension Headers {
  public override func isEqual(_ object: Any?) -> Bool {
    return (object as? Self)?.headers == self.headers
  }
}

// MARK: - CustomStringConvertible

extension Headers {
  public override var description: String {
    return "\(type(of: self)) \(self.headers.description)"
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/Headers.kt
package io.envoyproxy.envoymobile

/**
 * Base class that is used to represent header/trailer data structures.
 * To instantiate new instances, see `{Request|Response}HeadersBuilder`.
 */
open class Headers {
  internal val container: HeadersContainer

  /**
   * Internal constructor used by builders.
   *
   * @param container: The headers container to set.
   */
  internal constructor(container: HeadersContainer) {
    this.container = container
  }

  /**
   * Get the value for the provided header name. It's discouraged
   * to use this dictionary for equality key-based lookups as this
   * may lead to issues with headers that do not follow expected
   * casing i.e., "Content-Length" instead of "content-length".
   *
   * @param name: Header name for which to get the current value.
   *
   * @return The current headers specified for the provided name.
   */
  fun value(name: String): List<String>? {
    return container.value(name)
  }

  /**
   * Accessor for all underlying headers as a map.
   *
   * @return The underlying headers.
   */
  fun caseSensitiveHeaders(): Map<String, List<String>> {
    return container.caseSensitiveHeaders()
  }
}
